daffodil-core/src/main/scala/org/apache/daffodil/compiler/Compiler.scala

```diff
@@ -283,7 +284,31 @@ class Compiler private (var validateDFDLSchemas: Boolean,
 
   def reload(savedParser: java.nio.channels.ReadableByteChannel): DFDL.DataProcessor = {
     try {
-      val objInput = new ObjectInputStream(new GZIPInputStream(Channels.newInputStream(savedParser))) {
+      val is = Channels.newInputStream(savedParser)
+
+      // Read the magic number and version information for this saved parser
+      // directly from the input stream. That information is not compressed or
+      // serialized
+      val magicNumber = "DAFFODIL "
+      magicNumber.foreach { c =>
+        if (is.read() != c.toInt) throw new InvalidParserException("The saved parser is only compatible with an older version of Daffodil")
+      }
+
+      val sb = new StringBuilder()
+      var byte: Int = -1
+      while ({ byte = is.read(); byte > 0 }) {
+        sb.append(byte.toChar)
+      }
+      if (byte == -1) {
+        throw new InvalidParserException("The saved parser file is not the correct format")
+      }
+      if (sb.toString != Misc.getDaffodilVersion) {
+        throw new InvalidParserException("The saved parser is only compatible with Daffodil " + sb.toString)
```

Comment:
Add the current daffodil version that you are attempting to reload into here, to make the message more useful.